### DISCUSSION TITLE
State management

### DISCUSSION THREAD

**Discussion by jmarcgit** [mikejac/node-red-contrib-google-smarthome#618] [General]
Hi,

I try to automate my entry gate. I'm having issues with the state management.

I defined a gate type device and connected input and output to MQTT topics. One topic is used to send commands (switch) and the second one to report state (input sensor). Everything almost work except that the state of the gate in Google home is toggled automatically. I want to use a sensor to report the state of the gate because it can be opened with remotes as well. And it's nice to have continuous report of the state of an entry door, especially when I'm away.

![Screenshot 2024-12-05 23 16 28](https://github.com/user-attachments/assets/f319d82d-dd1d-4e84-aa9e-f69ca4fc4ddf)

So my question is : is there any way to fully control the state of an OpenClose trait ?

**Comment by Caprico85** [comment]
Sorry, but I don't understand the problem yet.

You mean, if somebody opens or closes the door with a remote (not via Google), you send the new state of the door into the Google device, but the new state is not shown in the app?

**Comment by jmarcgit** [comment]
No. When I send a new state it is show in the app.

The problem is that when I close the door using the Home app, the state moves immediately to close without waiting for the sensor event.

After analysing the debug log I think I have to change the state during the command processing to override the state returned  in the response to Google right ?

**Comment by Caprico85** [comment]
Thanks for the clarification.

Something similar was discussed in #474. For each command (such as "Unlock the door") Google waits for confirmation from our side. Currently, we confirm each command as successful, which leads to the state change in the app. Google also allows different return codes such as "Door not unlocked, motor is stuck". But this is currently not implemented in our module and there is no schedule on when (or if) it will be done.

> After analysing the debug log I think I have to change the state during the command processing to override the state returned in the response to Google right ?

After we handled a command, we immediately report the new state back to Google. Technically, this is not necessary, as our state would not be that different from the state Google sent us a few milliseconds ago. But that is how it's currently implemented. However, you cannot change this state that is reported immediately after the command (at least not without changing the source code). You can always send another new state report with the correct state right that, though.

**Reply by jmarcgit** [reply]
Thank you @Caprico85

1- I think that the confirmation can be sent a bit later. Maybe it could be possible to wait that the flow has completed and send the confirm message then. So if expection occurs, an error status could be returned.

2- Regarding the state, I think we could just manage an option for that. By default, state could be updated and reported automatically because this suits simple use cases where no sensors are available (current behaviour). But we could also offer the option to not updating the state automatically. Therefore state would be driven by sensors only.

What do you think ? Proposal 2 seems quite simple right ?

**Comment by Caprico85** [comment]
No, proposal 2 is not that simple. It is not our state reporting that changes the state display in the app.

The process is as follows:
1. User presses the "unlock door" button in the app
2. Google calculates that the new state is now "door unlocked". New state is not yet displayed in the app.
3. Google sends the command and the new state to node-red-contrib-google-smarthome.
4. We respond with "Command successful".
5. Google displays the new state in the app.
6. We send the state "door unlocked" to Google.
7. This changes nothing. Google already knew that the door was unlocked. This was the state that Google sent us right before.

As you see, Google displays the new state with step 5. This happens before we even reported our state in step 6. Disabling step and 6 and 7 with an option would not change anything, the new state is already visible in the app with step 5.

With the changes discussed in #474, this would change to:
1. User presses the "unlock door" button in the app (exactly as before)
2. Google calculates that the new state is now "door unlocked". New state is not yet displayed in the app. (exactly as before)
3. Google sends the command and the new state to node-red-contrib-google-smarthome. (exactly as before)
4. **Wait until we know more (e.g. got a sensor event), then respond with "Command successful" or "Command failed"**
5. If we sent "Command successful", Google displays the new state in the app.\
    If we sent "Command failed", Google discards the new state and shows an error message.
6. and 7.: report the state again (or not)

With this changes it would be possible to do what you want. You can delay the response until you got an answer from your sensor.
But as said, nobody has worked on #474 yet.

**Reply by jmarcgit** [reply]
I think I got the point. I didn't believe that Google Home was waiting the the reponse before refreshing the UI, because it was so fast... What is strange is that I set a 10 seconds blocking sleep in the device_exec function but the status is still updated immediately in Google Home.

**Reply by Caprico85** [reply]
You are right. I tried it myself now. The state immediately changes in the app. Even with a sleep it immediately changes. And even if we return an error code and the old state for the EXECUTE request, the state in the app changes and doesn't change back.

That's really strange. I have absolutely no idea how we could fix this.

**Reply by jmarcgit** [reply]
@Caprico85 : Update : I changed code to allow commands not to update state. It works. Button falls back to the current state after 1 second. I think we can add an option like "Automatically update state" in the node configuration that could be activated by default ? Do you want me to do the changes and create a pull request for that ?

**Reply by jmarcgit** [reply]
Hello,

I patched the code in order to disable the automatic state update.

I suggest to add a checkbox (activated by default) that will make the device state updated automatically like we have today. Uncheck the checkbox and then the state won't change automatically and must be updated from nodered/sensors.

Because it depends of the use case : when switching on a light, it doesn't matter much if you don't have a feedback. When opening a gate or a garage door you want to have a feedback.

Does it make sense ?

**Reply by jmarcgit** [reply]
Hello,

I'm still running the patched code but I would like to upgrade to the latest version of this project.

I just found something interesting. The state doesn't matter. If we want to support asynchronous processing of commands it's very simple to achieve it. The command action must return a PENDING status instead of SUCCESS. That's it. Doing this the button state won't be updated in Google Home.

The Node Red flow can store the request-id and send a notification message to Google to terminate the command when the gate sensor is activated. I don't know if we can create a such notification message with the current  state update feature. It looks like the code supports (it can process states or notifications)  it but I never tried it from a node.
More details are available here : https://developers.home.google.com/cloud-to-cloud/guides/gate#response

For me all is needed is to add a boolean "Async Command" setting at the device or trait level and just return a "PENDING" status when async mode is selected. @Caprico85 Do you think this feature could be implemented quickly ? I can test for you. I could do it by myself and do a PR but I think the PR review will take more time than implementing it.

The status trick explains why we always have seen a state change whatever we do. State can be set to the target value, so it's nice because it doesn't impact the existing devices logic. The PENDING status just says to Google to wait for the definive status that will SUCCESS or FAILURE. Let's see a realistic use case.
-You close a gate
-Node red returns the PENDING status and triggers the gate motor
-Node red store the identifier and waits for 1 minute
-If the sensor is triggered and if there is an identifier available then send a SUCCESS notification
-If nothing happens after a minute send a FAILURE notification and the current state

**Reply by jmarcgit** [reply]
@Caprico85 I did some tests and I was able to make the async process work. But there are a few caveats :

1. Only really works with Google speakers (not in the mobile app)
2. For mobile app the best option is to return the current state of the device and transition the state later, when we get feedbacks from the sensor. It doesn't hurt the speaker operations. When touching the button, the state visually changes during a couple of seconds and then displays the current state. Most important is that potential notifications are not triggered : It's just a visual glitch. Annoying but ok. Maybe one day Google will improve the user experience of the app.
3. If the operation is failing there is no notification in the mobile app (shame). But you can still set notifications on state changes to have some feedback on the mobile. Speaker announces the failure.
4. The main flow which receives the command can wait for a while and can trigger a **FAILURE** notification message if no event was received from the sensor within a reasonnable delay. As said before, that error is reported by Google speakers but not in the mobile app.
5. The secondary flow (sensor events) sends a **SUCCESS** notification message if a followup token is pending and then update device states
6. I use a global variable to share the followup token between the 2 flows

Impementation details:

1. Add the followup / async config option for the 3 traits supporting it : OpenClose, LockUnlock and NetworkControl
2. Add **notificationSupportedByAgent: true** to the **SYNC** replies when the followup option is active
4. When the followup is active, never report state immediately (_execDevice in HttpActions.ts)
5. When the followup is active, disable the state transition (updated function of Device.ts)
7. When the followup is active, return a **PENDING** status (_execDevice in HttpActions.ts)
